src/librustc_trans/trans/cleanup.rs

```diff
@@ -826,13 +832,15 @@ impl<'blk, 'tcx> CleanupHelperMethods<'blk, 'tcx> for FunctionContext<'blk, 'tcx
                 let bcx_in = self.new_block(&name[..], None);
                 let exit_label = label.start(bcx_in);
                 let mut bcx_out = bcx_in;
-                for cleanup in scope.cleanups.iter().rev() {
+                for cleanup in scope.cleanups.iter().rev().take(scope.cleanups.len() - skip) {
                     bcx_out = cleanup.trans(bcx_out, scope.debug_loc);
                 }
+                skip = 0;
                 exit_label.branch(bcx_out, prev_llbb);
                 prev_llbb = bcx_in.llbb;
 
-                scope.add_cached_early_exit(exit_label, prev_llbb);
+                let len = scope.cleanups.len();
```

Comment:
Nit: move this a little bit up and reuse `len` inside `take(scope.cleanups.len() - skip)`.